### PR TITLE
Dynamic data display name

### LINKED ISSUE #351: Custom display name for DynamicDataAttribute
## Description

Is it possible to add an option to the `DynamicDataAttribute` allowing similar functionality to the `DataRowAttribute.DisplayName`. This would allow tests with similar parameters to be given more meaningful names and therefore make it easier to determine which set of parameter is producing a particular result. 

A possible solution would be to set the a `DisplayName` property to the name of a method, just as `dynamicDataSourceName` is currently, and match the method signature of the default `GetDisplayName` method. This will allow a custom display name to be generated depending on the parameter set.  

## Current Behaviour

With the test class:

    [TestClass]
    public class Tests
    {
        [TestMethod]
        [DynamicData(nameof(Tests.TestMethodTests))]
        public void TestMethod(string param1, string param2)
        {
            Assert.IsNotNull(param1);
        }

        private static IEnumerable<object[]> TestMethodTests
        {
            get
            {
                yield return new object[] { "", null };
                yield return new object[] { null, "" };
            }
        }
    }

Displays output where it is not clear which set of parameters caused the failure

    Failed   TestMethod (,)
    Error Message:
        Assert.IsNotNull failed.
    Stack Trace:
        ...

## Possible Solution

    // ....
        [DynamicData(nameof(Tests.TestMethodTests), DisplayName = nameof(Tests.GetTestMethodDisplayName))]
    // ...
        public static string GetTestMethodDisplayName(MethodInfo methodInfo, object[] data)
        {
            // ....
        }
    // ....

## Current Workaround

As a work around I have been adding an unused sting parameter to identify the test e.g:

    // ....
        public void TestMethod(string testName, string param1, string param2)
    // ....
            yield return new object[] { "Param2 null", "", null };
            yield return new object[] { "Param1 null", null, "" };
    // ....

### PR DESCRIPTION
**Pull request by bstoney**
Fixes #351 
Added dynamic display name to DynamicDataAttribute

### CODE REVIEW

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DataRowAttributeTests.cs [hunk 1]
```diff
@@ -81,5 +81,20 @@ public void GetDisplayNameShouldReturnAppropriateName()
             displayName = dataRowAttribute.GetDisplayName(this.testMethodInfo, data2);
             Assert.AreEqual("DataRowTestMethod (First,,Second)", displayName);
         }
+
+        [TestMethod]
+        public void GetDisplayNameShouldReturnAppropriateNameWithDataRowDisplayName()
```
Comment (jayaranigarg) [inline test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DataRowAttributeTests.cs]: Please consider changing testmethod name to GetDisplayNameShouldReturnProvidedDisplayName() or GetDisplayNameShouldReturnSpecifiedDisplayName()

File: src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs [hunk 2]
```diff
@@ -70,6 +70,16 @@ public DynamicDataAttribute(string dynamicDataSourceName, Type dynamicDataDeclar
             this.dynamicDataDeclaringType = dynamicDataDeclaringType;
         }
 
+        /// <summary>
+        /// Gets or sets the name of method used to customize the display name in test results.
+        /// </summary>
+        public string DynamicDisplayName { get; set; }
```
Comment (jayaranigarg) [inline src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs]: Consider renaming this to "Dynamic**Data**DisplayName"

File: src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs [hunk 3]
```diff
@@ -70,6 +70,16 @@ public DynamicDataAttribute(string dynamicDataSourceName, Type dynamicDataDeclar
...
+
+        /// <summary>
+        /// Gets or sets the declaring type used to customize the display name in test results.
+        /// </summary>
+        public Type DynamicDisplayNameDeclaringType { get; set; }
```
Comment (jayaranigarg) [inline src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs]: Same here. "Dynamic**Data**DisplayNameDeclaringType"

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs [hunk 4]
```diff
@@ -60,7 +60,7 @@ public void GetDataShouldReadDataFromProperty()
         public void GetDataShouldReadDataFromPropertyInDifferntClass()
         {
             var methodInfo = this.dummyTestClass.GetType().GetTypeInfo().GetDeclaredMethod("TestMethod1");
-            this.dynamicDataAttribute = new DynamicDataAttribute("ReusableTestDataProperty", typeof(DummyTestClass));
+            this.dynamicDataAttribute = new DynamicDataAttribute("ReusableTestDataProperty2", typeof(DummyTestClass2));
```
Comment (jayaranigarg) [inline test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs:63]: Nice catch 👍

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs [hunk 5]
```diff
@@ -121,6 +121,41 @@ public void GetDisplayNameShouldReturnDisplayName()
...
+            var data = new object[] { 1, 2, 3 };
+
+            this.dynamicDataAttribute.DynamicDisplayName = "GetCustomDynamicDataDisplayName2";
+            this.dynamicDataAttribute.DynamicDisplayNameDeclaringType = typeof(DummyTestClass2);
+            var displayName = this.dynamicDataAttribute.GetDisplayName(this.testMethodInfo, data);
+            Assert.AreEqual("DynamicDataTestWithDisplayName TestMethod1 with 3 parameters", displayName);
+        }
+
+        [TestFrameworkV1.TestMethod]
+        public void GetDisplayNameShouldThrowExceptionIfMethodDoesNotMatchExpectedSignature()
```
Comment (jayaranigarg) [inline test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs]: Please add Tests for cases like : 
1. Method not existing i.e InvalidCustomDynamicDataDisplayName is not present in class
2. Method has wrong signature i.e. when first parameter doesn't match, second parameter doesn't match, return type doesn't match. Write tests for each of them separately.

File: src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx [hunk 6]
```diff
@@ -283,4 +283,7 @@ Stack Trace: {4}</value>
   <data name="DynamicDataValueNull" xml:space="preserve">
     <value>Value returned by property or method {0} shouldn't be null.</value>
   </data>
+  <data name="DynamicDataDisplayName" xml:space="preserve">
```
Comment (jayaranigarg) [inline src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx:286]: Also please run "build.cmd -uxlf" so the string addition is localizable. You would have to commit the xlf files changes post running the script.

**Comment by jayaranigarg** [comment]
@bstoney : Apologies for the delay. 
We have reviewed you PR. Kindly do the needful changes so that we can push this in.

**Comment by jayaranigarg** [comment]
@bstoney : Can you make the necessary changes here please?

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DataRowAttributeTests.cs (hunk 1 above)
Comment (bstoney) [reply test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DataRowAttributeTests.cs]: Done

File: src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs (hunk 2 above)
Comment (bstoney) [reply src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs]: Done

File: src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs (hunk 3 above)
Comment (bstoney) [reply src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs]: Done

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs (hunk 4 above)
Comment (bstoney) [reply test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs:63]: Cheers 😄

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs (hunk 5 above)
Comment (bstoney) [reply test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs]: Done. Also picked up a bug where I wasn't checking that the method was static.

File: src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx (hunk 6 above)
Comment (bstoney) [reply src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx:286]: I ran this and have committed the xlf files. I'm assuming someone or some magic will do the translation?

File: test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs (hunk 5 above)
Comment (jayaranigarg) [reply test/UnitTests/MSTest.Core.Unit.Tests/Attributes/DynamicDataAttributeTests.cs]: Awesome 👍

File: src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx (hunk 6 above)
Comment (jayaranigarg) [reply src/TestFramework/MSTest.Core/Resources/FrameworkMessages.resx:286]: Yes please. We just needed modified xlf files. We will take care of the translation 😄

File: src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs [hunk 7]
```diff
@@ -131,7 +141,34 @@ public IEnumerable<object[]> GetData(MethodInfo methodInfo)
...
+                if (parameters.Length != 2 ||
+                    parameters[0].ParameterType != typeof(MethodInfo) ||
+                    parameters[1].ParameterType != typeof(object[]) ||
+                    method.ReturnType != typeof(string) ||
+                    !method.IsStatic)
+                {
+                    throw new ArgumentNullException(
+                        string.Format(
+                            FrameworkMessages.DynamicDataDisplayName,
+                            this.DynamicDataDisplayName,
```
Comment (jayaranigarg) [inline src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs:165]: @bstoney : Method signature should also have "public" access modifier specified. Can you please add that as well??

Comment (jayaranigarg) [reply src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs:165]: You will have to update xlf files as well if you modify  FrameworkMessages.DynamicDataDisplayName string.

Comment (bstoney) [reply src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs:165]: Done. Interestingly it did work with a private static method even though the documentation for `GetDeclaredMethod` states that "Returns an object that represents the specified **public** method declared by the current type." I have updated the exception message and added a test for public accessibility.

Comment (jayaranigarg) [reply src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs:165]: Super 👍

**Comment by smadala** [comment]
@bstoney Thanks for contribution 👍 . Can you please update documentation as well [here](https://github.com/Microsoft/testfx-docs/blob/master/RFCs/006-DynamicData-Attribute.md) ? then we can merge this PR.

**Pull request merged**